internal/installercache/installercache.go

```diff
@@ -183,33 +291,34 @@ func (i *Installers) evict() {
 		return nil
 	}
 
-	err := filepath.Walk(i.cacheDir, visit)
+	err := filepath.Walk(i.config.CacheDir, visit)
 	if err != nil {
 		if !os.IsNotExist(err) { //ignore first invocation where the cacheDir does not exist
-			i.log.WithError(err).Errorf("release binary eviction failed to inspect directory %s", i.cacheDir)
+			i.log.WithError(err).Errorf("release binary eviction failed to inspect directory %s", i.config.CacheDir)
 		}
 		return
 	}
 
-	//prune the hard links just in case the deletion of resources
-	//in ignition.go did not succeeded as expected
-	for idx := 0; idx < len(links); idx++ {
-		finfo := links[idx]
-		//Allow a grace period of 5 minutes from the link creation time
-		//to ensure the link is not being used.
-		grace := time.Now().Add(DeleteGracePeriod).Unix()
-		if finfo.info.ModTime().Unix() < grace {
-			os.Remove(finfo.path)
-		}
-	}
+	// TODO: We might want to consider if we need to do this longer term, in theory every hardlink should be automatically freed.
+	// For now, moved to a function so that this can be tested in unit tests.
+	i.pruneExpiredHardLinks(links, linkPruningGracePeriod)
 
-	//delete the oldest file if necessary
-	for totalSize >= int64(float64(i.storageCapacity)*CacheLimitThreshold) {
-		finfo, _ := files.Pop()
-		totalSize -= finfo.info.Size()
-		//remove the file
-		if err := i.evictFile(finfo.path); err != nil {
-			i.log.WithError(err).Errorf("failed to evict file %s", finfo.path)
+	// delete the oldest file if necessary
+	// prefer files without hardlinks first because
+	// 1: files without hardlinks are likely to be least recently used anyway
+	// 2: files without hardlinks will immediately free up storage
+	queues := i.splitQueueOnHardLinks(files)
+	for _, q := range queues {
+		for i.shouldEvict(totalSize) {
```

Comment:
Because we're evaluating this each loop it feels very much like we should not need the loop in get and should never hit the capacity error, right?

Doesn't this effectively ensure we'll evict releases until we have enough space to download a new one? Or am I misunderstanding something?

Reply:
Already discussed above, I think something has been misunderstood here.